File: 算法/排序/冒泡排序.cpp
//
// Created by Momo on 2023/9/12.
//
#include <iostream>

using namespace std;

template<class T>
void bubble_sort(vector<T> &v) {
    for (int i = 0; i < v.size() - 1; ++i) {
        bool swapped = false;
        for (int j = 0; j < v.size() - 1 - i; ++j)
            if (v.at(j) > v.at(j + 1)) {
                swap(v.at(j), v.at(j + 1));
                swapped = true;
            }
        // 如果冒泡了一轮，但是没有任何交换，我们可以提前退出
        if (swapped) break;
    }
}

template<class T>
void show_data(const vector<T> &v) {
    for (const auto &num: v)
        cout << num << " ";
    cout << endl;
}

int main() {
    vector<int> v{10, 2, 3, 12, 5, 9, 0};

    vector<string> vs{"you", "asshole", "bitch"};

    cout << "Original data: " << endl;
    show_data(v);

    bubble_sort(v);

    cout << "After Sorting: " << endl;
    show_data(v);

    return 0;
}

File: 算法/排序/冒泡排序.py
def bubble_sort(unsorted_ls: list):
    for i in range(len(unsorted_ls)-1):
        swapped: bool = False
        for j in range(i, len(unsorted_ls)-1-i):
            if unsorted_ls[j] > unsorted_ls[j + 1]:
                unsorted_ls[j], unsorted_ls[j + 1] = unsorted_ls[j + 1], unsorted_ls[j]
                swapped = True
            if swapped: break


if __name__ == '__main__':
    ls = [20, 10, 30, 40]
    bubble_sort(ls)
    for i in ls:
        print(f'{i} ', end='')
    print()